papyri/gen.py

```diff
@@ -1135,145 +1223,57 @@ def get_example_data(
-                                did_except,
-                            ]
-                    # we've executed, we now want to compare output
-                    # in the docstring with the one we produced.
-                    if (out == repr(res)) or (res is None and out == []):
-                        pass
-                    else:
-                        pass
-                        # captured output differ TBD
-                entries = parse_script(script, ns=ns, prev=acc, config=config, where=qa)
-                if entries is None:
-                    entries = [("jedi failed", "jedi failed")]
-                entries = _add_classes(entries)
-                tok_entries = [GenToken(*x) for x in entries]  # type: ignore
 
-                acc += "\n" + script
-                example_section_data.append(
-                    Code(tok_entries, "\n".join(item.out), ce_status)
-                )
-                for figname, _ in figs:
-                    example_section_data.append(
-                        Fig(
-                            RefInfo.from_untrusted(
-                                self.root, self.version, "assets", figname
-                            )
-                        )
-                    )
-                all_figs.extend(figs)
+        sys_stdout = sys.stdout
+        def dbg(*args):
+            for arg in args:
+                sys_stdout.write(f"{arg}\n")
+            sys_stdout.flush()
 
-        # TODO fix this if plt.close not called and still a ligering figure.
-        fig_managers = executor.fig_man()
+        try:
+            filename = inspect.getfile(obj)
+        except TypeError:
+            filename = None
+        try:
+            lineno = inspect.getsourcelines(obj)[1]
+        except (TypeError, OSError):
+            lineno = None
+
+        doctest_runner = PapyriDocTestRunner(gen=self, obj=obj, qa=qa,
+                                             config=config,
+                                             # TODO: Make optionflags configurable
+                                             optionflags=doctest.ELLIPSIS)
+        example_section_data = Section([], None)
+        def debugprint(*args):
+            """
+            version of print that capture current stdout to use during testing to debug
+            """
+            sys_stdout.write(" ".join(str(x) for x in args) + "\n")
+
+        for block in blocks:
+            doctests = doctest.DocTestParser().get_doctest(block,
+                                                           doctest_runner.globs,
+                                                           obj.__name__, filename,
+                                                           lineno)
+            if doctests.examples:
+                doctest_runner.run(doctests, out=debugprint)
+                example_section_data.extend(doctest_runner.example_section_data)
+            else:
+                example_section_data.append(MText(block))
```

Comment:
I'm not sure if the disposition of newlines is correct here. We may need to add a `MText('\n')` at the end here or something like that.

Reply:
that is possible, sorry about that.

Reply:
I thin we also need to create the doctest runner outside of the loop, and maybe concatenate blocks/doctests, as many examples in the scipy ecosystem seem to have test bettween code examples that are supposed to keep the current namespace. 

See https://github.com/jupyter/papyri/blob/63a5ea83e4c0cd4f5a80413fc57078773c8bdec0/papyri/examples.py which is a copy of SciPy univariate spline, whcih fails which a bunch of name error.

```
papyri gen examples/papyri.toml --only papyri.examples:example2
```

Reply:
Here's what that example generates:

```json
  "example_section_data": {
    "children": [
      {
        "type": "code",
        "value": "import numpy as np\n"
      },
      {
        "type": "code",
        "value": "import matplotlib.pyplot as plt\n"
      },
      {
        "type": "code",
        "value": "from scipy.interpolate import UnivariateSpline\n"
      },
      {
        "type": "code",
        "value": "import numpy as np\n"
      },
      {
        "type": "code",
        "value": "import matplotlib.pyplot as plt\n"
      },
      {
        "type": "code",
        "value": "from scipy.interpolate import UnivariateSpline\n"
      },
      {
        "type": "code",
        "value": "x = np.linspace(-3, 3, 50)\n"
      },
      {
        "type": "code",
        "value": "y = np.exp(-(x ** 2)) + 0.1 * np.random.randn(50)\n"
      },
      {
        "type": "code",
        "value": "plt.plot(x, y, \"ro\", ms=5)\n"
      },
      {
        "type": "text",
        "value": "Use the default value for the smoothing parameter:"
      },
      {
        "type": "code",
        "value": "import numpy as np\n"
      },
      {
        "type": "code",
        "value": "import matplotlib.pyplot as plt\n"
      },
      {
        "type": "code",
        "value": "from scipy.interpolate import UnivariateSpline\n"
      },
      {
        "type": "code",
        "value": "x = np.linspace(-3, 3, 50)\n"
      },
      {
        "type": "code",
        "value": "y = np.exp(-(x ** 2)) + 0.1 * np.random.randn(50)\n"
      },
      {
        "type": "code",
        "value": "plt.plot(x, y, \"ro\", ms=5)\n"
      },
      {
        "type": "code",
        "value": "spl = UnivariateSpline(x, y)\n"
      },
      {
        "type": "code",
        "value": "xs = np.linspace(-3, 3, 1000)\n"
      },
      {
        "type": "code",
        "value": "plt.plot(xs, spl(xs), \"g\", lw=3)\n"
      },
      {
        "type": "text",
        "value": "Manually change the amount of smoothing:"
      },
      {
        "type": "code",
        "value": "import numpy as np\n"
      },
      {
        "type": "code",
        "value": "import matplotlib.pyplot as plt\n"
      },
      {
        "type": "code",
        "value": "from scipy.interpolate import UnivariateSpline\n"
      },
      {
        "type": "code",
        "value": "x = np.linspace(-3, 3, 50)\n"
      },
      {
        "type": "code",
        "value": "y = np.exp(-(x ** 2)) + 0.1 * np.random.randn(50)\n"
      },
      {
        "type": "code",
        "value": "plt.plot(x, y, \"ro\", ms=5)\n"
      },
      {
        "type": "code",
        "value": "spl = UnivariateSpline(x, y)\n"
      },
      {
        "type": "code",
        "value": "xs = np.linspace(-3, 3, 1000)\n"
      },
      {
        "type": "code",
        "value": "plt.plot(xs, spl(xs), \"g\", lw=3)\n"
      },
      {
        "type": "code",
        "value": "spl.set_smoothing_factor(0.5)\n"
      },
      {
        "type": "code",
        "value": "plt.plot(xs, spl(xs), \"b\", lw=3)\n"
      },
      {
        "type": "code",
        "value": "plt.show()\n"
      },
      {
        "type": "Fig",
        "value": {
          "kind": "assets",
          "module": "papyri",
          "path": "fig-papyri.examples:example2-0-19992817.png",
          "type": "RefInfo",
          "version": "0.0.8"
        }
      }
    ],
```

The `code` and `text` elements are separated. The main thing I'm not sure about, as I said, is whether there are the correct number of `\n`s for rendering, since the text is actually separated from the code by two newlines.

Reply:
The \n thing should be fixed with my latest changes.